src/compiler/types.ts

```diff
@@ -345,17 +349,31 @@ module ts {
         // error.
         ThisNodeHasError = 1 << 5,
 
+        // If this node was parsed in the parameters of an async function.
+        AsyncParameter = 1 << 6,
+
+        // If this node was parsed in the 'await' context created when parsing an async function.
+        Await = 1 << 7,
+
         // Context flags set directly by the parser.
-        ParserGeneratedFlags = StrictMode | DisallowIn | Yield | GeneratorParameter | Decorator | ThisNodeHasError,
+        ParserGeneratedFlags = StrictMode | DisallowIn | Yield | GeneratorParameter | Decorator | ThisNodeHasError | AsyncParameter | Await,
 
+        // Context flags passed as part of the modified ES6 grammar.
+        YieldAndGeneratorParameterFlags = Yield | GeneratorParameter,
+        AwaitAndAsyncParameterFlags = Await | AsyncParameter,
+        AllParameterFlags = YieldAndGeneratorParameterFlags | AwaitAndAsyncParameterFlags,       
+        
         // Context flags computed by aggregating child flags upwards.
 
         // Used during incremental parsing to determine if this node or any of its children had an
         // error.  Computed only once and then cached.
-        ThisNodeOrAnySubNodesHasError = 1 << 6,
+        ThisNodeOrAnySubNodesHasError = 1 << 8,
+
+        // Used to know if we've computed whether any children of this node are or contain an 'await' or 'yield' expression.
+        ThisNodeOrAnySubNodesHasAwaitOrYield = 1 << 9,

```

Comment:
I don't this is to support the downlevel emit, once I've updated it to the latest from master. Almost everything in this PR except emitter.ts has already been reviewed in #1781. I can remove it for now, but I'll still end up needing it in the future.